### ISSUE TITLE
Recognize FTP files as MAF file when generating MafIO.MafIndex

### ISSUE THREAD

**Issue by asztandi** [biopython/biopython#4656]
### Improvement

When generating an index file for a MAF file (https://github.com/biopython/biopython/blob/master/Bio/AlignIO/MafIO.py#L263) it would be good if an FTP file was also recognized as an input MAF file. The application does not access to the local filesystem.

Thank you for considering.

**Comment by peterjc** [comment]
I don't understand your use case. If you don't have access to the local file system, where would you put the index file? Or are you thinking of having both the MAF file and the index side-by-side on the FTP server?

**Comment by asztandi** [comment]
Yes, that would be the ideal scenario, having the MAF file and the index side-by-side on the FTP server.

**Comment by peterjc** [comment]
The index is stored as an SQLite3 file, you might have to download all of it. The filenames within it are relative paths. You would need to convert that to an FTP URL (or in principle HTTPS URL) and open the remote MAF file with random access.

(I recall similar discussions with BAM files or similar)

This sounds _possible_, probably with a subclass of the local file index lookup in https://github.com/biopython/biopython/blob/master/Bio/File.py

Is it something you'd like to explore yourself?

**Comment by mdehoon** [comment]
Could you use a bigMaf file instead (i.e. a MAF file converted to a bigMaf file using either Biopython or UCSC's tools)? A bigMaf file has a built-in index, so then only one file is needed.

**Comment by asztandi** [comment]
Using a BigMaf file would be great. Maybe I missed it but I couldn't find a way to randomly search for a region using a Bigmaf file, something similar to using a MafIndex:
`idx = MafIO.MafIndex(index_file, maf_file, anchor_seq_name)
 alignments = idx.search([start], [end])`
Please let me know if one can randomly retrieve a region using a Bigmaf file.

The Bigmaf should also be readable from the FTP or from a cloudian storage (no access to local filesystem in this case either).

**Comment by asztandi** [comment]
Looked into working with a BigMaf file.
Both parsing  
```
alignments = BioAlign.parse(bigmaf_file, "bigmaf")
```
followed by searching  
```
sel_aligns = alignments.search(anchor_seq_name, start, end)
for a in sel_aligns:
```
is slower than using an already created MAF index and searching with it. 

The other issue is I can't seem to find a way to iterate through the actual sequence. The code below throws an exception with the message '_Sequence content is only partially defined_':
```
for a in sel_aligns:
        start, end = a.coordinates[0, 0], a.coordinates[0, -1]
        print (start, end)
        for seq in a.sequences:
            na =seq.seq     # variable value: Seq({26235490: 'CAAGGGCAAAGATTCACTTTCAACTCAGAATTCGGGTTGCCACTTGTAGGAGGA...GTC'}, length=57227415)
            na_str = str(na_str)     # throws exception
```
Overall, I'm still inclined to use MafIO.MafIndex. 

@peterjc:
>Is it something you'd like to explore yourself?

Yes, I would like to. Speed is very crucial for this project because it's a web application. 
I've looked into pybbi and bx-python and their BigMaf parsing/searching functions are much slower.

**Comment by mdehoon** [comment]
> The code below throws an exception with the message 'Sequence content is only partially defined':

This is because of this line:
```python
        for seq in a.sequences:
```
`Bio.Align.parse` returns `Alignment` objects, not `MultipleSeqAlignment` objects like `Biio.AlignIO`.
The sequences in `a.sequences` are the unaligned sequences, in which only the aligned sequence is defined. Try `print(repr(seq))` to see them in detail, and  `print(a)` to see the alignment.

> Both parsing followed by searching is slower than using an already created MAF index and searching with it.

Do you have some timings? Note that parsing has to be done only once.

**Comment by asztandi** [comment]
Thank you for the explanation regarding the exception.
> Do you have some timings? 

Yes, please find them below.  

Method A 
```
idx = MafIO.MafIndex(index_file, maf_file, anchor_seq_name)
alignments = idx.search([start], [end])
 ```
Method B
 ```
alignments = BioAlign.parse(bigmaf_file, "bigmaf")
sel_aligns = alignments.search(anchor_seq_name, start, end)
 ```
I measured it for a smaller MAF file (human chromosome Y)
size(MAF_file) = 222MB
size (Bigmaf_file) = 16MB
Most of the files I'm working with are > 15GB
The maf, the bigmaf and the mafindex file are all in the same network directory.

For _[start, end) = [340910, 341090]_  - range towards the beginning of the MAF file
Method A: 0.014s
Method B: 1.38s

For _[start, end) = [26235490, 26235690]_ - range towards the end of the MAF file
Method A: 0.017s
Method B: 1.41s

All measurements are performed on an AlmaLinux workstation.
Clearly, method A is much faster.

**Comment by mdehoon** [comment]
Thanks for the timings.
When I try this on the multiz100way multiple genome alignments from UCSC for chrY, I am getting 0.011 seconds for the bigMaf file (executed locally, not over the network). The search function alone takes 5e-5 seconds.
Do you get the same timings when you download the bigMaf file and try on the downloaded file?

**Comment by asztandi** [comment]
> When I try this on the multiz100way multiple genome alignments from UCSC for chrY, I am getting 0.011 seconds for the bigMaf file 

I see the exact same speed as you mention. 
I'm not using the multiZ alignment data from UCSC, because the aligned regions are very short (fragmented alignment) I'm using EPO data from Ensemble. I think the shortness of the aligned regions contribute to the fast search in the case of the multiZ data. Here I'm searching for a much smaller range ( < 2000bp) than the average alignment length, and it looks like the mafIndex does a better job in this case.

**Comment by mdehoon** [comment]
@asztandi  Would you be willing to share your bigMaf file?

**Comment by asztandi** [comment]
Yes, please take a look at: https://ftp.ncbi.nlm.nih.gov/comparative-genome-viewer/mcgv/43_mammals_epo/filtered/
The bigMaf is the 43_mammals.epo.1_filtered.bb This is for chromosome 1. 
The MAF file for chr Y is 43_mammals.epo.Y_filtered.maf

**Comment by mdehoon** [comment]
Thanks for sharing the files.

Most of the slowdown was due to large data chunks being copied while parsing. This is fixed by #4660 .

To speed up the search, when creating the bigMaf file, you can use the `-unc` argument to `bedToBigBed` to generate an uncompressed file, and `-itemsPerSlot=32` to create a deeper search tree.

With these three changes, I am getting a search time of 0.6 seconds.

This is still much longer than `MafIO.MafIndex`. Most of the time goes into converting the MAF alignment block to alignment coordinates (which are used internally to represent the sequence alignment).

So my question now is: If you search for a specific region and get the alignment object, what will you do with it? Do you just want to print out the MAF alignment block, or get the genome coordinates for the other species, or something else?

**Comment by asztandi** [comment]
Thank you so much for all your suggestions. We will incorporate them in our bigMaf workflow.

After we search for a specific region (that may be within one MAF block or may span multiple MAF blocks) we would want to have full access to all sequences within that region: we would like to print out the sequences, count matches/mismatches/gaps relative to the anchor sequence and export the region into a new FASTA alignment.

Meanwhile, we solved the issue of having access to the filesystem when creating/reading the MafIndex files. So my original question has been resolved.  However, the idea of using one file per chromosome is neater and less error prone.

**Comment by mdehoon** [comment]
This branch:

https://github.com/mdehoon/biopython/tree/align_coordinates

has a C implementation of the most time-consuming parts of the bigMaf parser. 

With Method A:
```python
>>> idx = MafIO.MafIndex(index_file, maf_file, anchor_seq_name)
>>> alignments = idx.search([start], [end])
>>> alignment = next(alignments)
```
I am getting
- 420 microseconds for reading the index
- 3 microseconds for searching
- 12644 microseconds for pulling out one alignment.

With Method B:
```python
>>> all_alignments = Align.parse(maf_file, "bigMaf")
>>> alignments = all_alignments.search([start], [end])
>>> alignment = next(alignments)
```
I am getting
- 1583 microseconds for reading the indexed bigMaf file
- 1 microsecond for searching
- 30140 microseconds for pulling out one alignment.
so about 2.5 times slower than Method A.

However, Method A generates `MultipleSeqAlignment` objects, while Method B generates `Alignment` objects, and internally these classes store the alignment in very different ways. It would be interesting to find out how much time your subsequent analysis steps take for Method A and B, which of course depends on what exactly you want to do with the alignment.

**Comment by asztandi** [comment]
Thank you @mdehoon for all the timing details!
I also noticed, as you mentioned in your comment, that obtaining one alignment is much slower using Method B. That was another reason of favoring Method A over Method B.
